pennylane/templates/broadcast.py

```diff
@@ -13,49 +13,59 @@
 # limitations under the License.
 r"""
 Contains the ``broadcast`` template constructor.
+
+To add a new structure, extend the variables ``OPTIONS``, ``n_parameters`` and ``wire_sequence``, and
+update the list of parameter numbers in the docstring. Optionally add a usage example at the end of the
+``UsageDetails`` directive.
 """
 # pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
 from collections import Iterable
 
 from pennylane.templates.decorator import template
-from pennylane.templates.utils import _check_wires, _check_type, _get_shape
+from pennylane.templates.utils import _check_wires, _check_type, _get_shape, _check_is_in_options
 
 
 @template
-def Broadcast(block, wires, parameters=None, kwargs={}):
-    r"""Applies a (potentially parametrized) single-qubit unitary ``block`` to each wire.
+def broadcast(block, wires, structure="single", parameters=None, kwargs={}):
```

Comment:
On the first note, I got `ValueError: CNOT: wrong number of wires. 1 wires given, 2 expected.`, when calling `broadcast(qml.CNOT, wires=[0,1,2,3])` which is expected right now.

Reply:
About ``structure=None``, I thought about it, but I think it is by no means clear what the default is here. I want to force the user to fix a structure each time. So if you don't mind, I would leave it this way!

Reply:
About ``kwargs=None``, thanks, that makes total sense! I didn't know it is even a best practice.

Reply:
Sure! Would you say that the default for `structure` is unclear since in the `double` case there could be `odd` or `even` cases?

Reply:
Well, I'd say it is unclear that a user considers ``double`` as a default, and the gate-wire test makes it very verbose... On the other hand, you are right that ``broadcast`` suggests ``single`` and ``double`` as defaults.

We could easily (without feature breaking) introduce a default later, what do you think?

Reply:
Sorry, @antalszava, reading through this again your idea makes total sense, since we actually *have* a default `single`, which breaks for two-qubit gates. I understand your point now. Will implement your suggested change!

Reply:
The only problem is that templates have no attribute ``num_wires``

Reply:
Shall we just make ``structure`` a positional argument?

Reply:
Yes, I think it might be worth making `structure` required without any defaults for now.

Making it `structure=single`  to have a default value seems to imply the following cases:
* The user specifies `structure` (most preferable case, though requires user's attention) :heavy_check_mark: 
* The user does not specify `structure`, but either indeed a single-qubit template was passed as a block and the default value is valid :heavy_check_mark: 
* The user does not specify `structure`, but passes a PennyLane native operation, in which case `structure` can be deduced (e.g. double or so) :heavy_check_mark: 
* The user does not specify `structure` and passes a custom template which is not single-qubit, in which case an error would be raised :heavy_multiplication_x: 

It would be really nice to have some sort of addition perhaps to the `@template` decorator or in other way imposing the constraint onto custom templates that they need to specify the `num_wires`. But this is entirely out of scope for this PR.